### PR TITLE
fix: Dialect requires derived table alias

### LINKED ISSUE #12993: MySQL does not support derived tables without aliases
### Describe the bug

In some logical plans, `.derive()` creates a new subquery as a derived table like `FROM (SELECT ...)`.

In MySQL, every derived table requires an alias. `.derive()` does not supply an alias to these derived tables, so the queries fail in MySQL.

### To Reproduce

Run a query on MySQL that triggers a logical plan derive, like:

```
select min(ta.j1_id) as j1_min from j1 ta order by min(ta.j1_id) limit 10;
```

Observe the query fail, because the re-written SQL contains a derived table without an alias.

### Expected behavior

The query should succeed without a syntax error.

### Additional context

As seen from the MySQL manual: https://dev.mysql.com/doc/refman/8.4/en/derived-tables.html

> The [AS] tbl_name clause is mandatory because every table in a FROM clause must have a name.

### PR DESCRIPTION
**Pull request by peasee**
## Which issue does this PR close?

Closes #12993.

## Rationale for this change

Fixes a bug preventing derived logical plans from running in MySQL due to a requirement that every derived table must have an alias:

> The [AS] tbl_name clause is mandatory because every table in a FROM clause must have a name.

## What changes are included in this PR?

Includes a dialect update to specify whether derived tables require aliases. For dialects that do, set a static alias. I've simply selected the name of the operation prefixed with `derived_`, like `derived_sort`, etc.

## Are these changes tested?

Yes

## Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/sql/tests/cases/plan_to_sql.rs
Comment (peasee) [reply datafusion/sql/tests/cases/plan_to_sql.rs:269]: I added another test case for a more complex SQL that results in 2 nested derives for `derived_sort` and `derived_distinct` in the same query.

I'm not too sure what SQL to use to trigger `derived_limit` or `derived_projection` though... any ideas that I could include?

Comment (goldmedal) [reply datafusion/sql/tests/cases/plan_to_sql.rs:269]: For `derived_projection`, I found the SQL can trigger:
```sql
select j1_id from (select 1 as j1_id)
 -> SELECT `j1_id` FROM (SELECT 1 AS `j1_id`) AS `derived_projection`
```
However, I guess it's not a valid SQL for MySQL, right? I tried it in DataFusion and DuckDB, they accept it but Postgres doesn't allow it. Maybe, it could be a `DataFusion` dialect to `MySQL` dialect case.

For `derived_limit`, a similar case can trigger it:
```sql
select * from (select * from j1 limit 10) 
  -> SELECT * FROM (SELECT * FROM `j1` LIMIT 10) AS `derived_limit`
```

**Comment by goldmedal** [comment]
By the way, I tried to pull your branch and run the tests on my laptop but the tests always fail. After merging with the latest main branch, the tests are passed.

File: datafusion/sql/tests/cases/plan_to_sql.rs
Comment (peasee) [reply datafusion/sql/tests/cases/plan_to_sql.rs:269]: It is valid for MySQL, I just tested it out in my CLI. I've added both as new cases, thanks!

**Review by goldmedal — APPROVED** [review]
It's a good improvement for the derived table. 👍 It makes sense to me.
I noticed some databases don't support this kind of unnamed subquery, see https://github.com/apache/datafusion/pull/12896#issuecomment-2408609989
This PR would be helpful.

**Comment by goldmedal** [comment]
Maybe @phillipleblanc and @sgrebnov want to take a look at this PR.

**Review by phillipleblanc — APPROVED** [review]

**Comment by phillipleblanc** [comment]
Looks good to me, thanks @goldmedal and @peasee

**Comment by alamb** [comment]
@goldmedal  perhaps you would like to merge this PR as a way to verify your permissions are configured correctly?

**Comment by goldmedal** [comment]
> @goldmedal perhaps you would like to merge this PR as a way to verify your permissions are configured correctly?

Sure, I plan to merge this PR if no more comments tomorrow.

**Comment by peasee** [comment]
> > @goldmedal perhaps you would like to merge this PR as a way to verify your permissions are configured correctly?
> 
> Sure, I plan to merge this PR if no more comments tomorrow.

Let's hold off on merging this for now. I think I've introduced a regression that I'd like to test more first before merging.

**Comment by peasee** [comment]
> > > @goldmedal perhaps you would like to merge this PR as a way to verify your permissions are configured correctly?
> > 
> > 
> > Sure, I plan to merge this PR if no more comments tomorrow.
> 
> Let's hold off on merging this for now. I think I've introduced a regression that I'd like to test more first before merging.

Okay, looks like my PR didn't introduce the regression but there's a somewhat related bug currently on `main`. Take the following SQL:

```
SELECT j1_id FROM (SELECT ta.j1_id AS j1_id FROM j1 ta);
```

This gets rewritten in the `GenericDialect` to:

```
SELECT ta.j1_id FROM (SELECT ta.j1_id FROM j1 AS ta)
```

Which is invalid, because the subquery is un-aliased. I tested it in DuckDB to confirm:

```
D SELECT ta.j1_id FROM (SELECT ta.j1_id FROM j1 ta);
Binder Error: Referenced table "ta" not found!
Candidate tables: "unnamed_subquery"
LINE 1: SELECT ta.j1_id FROM (SELECT ta.j1_id FROM j1 ...
```

Let's merge my PR, and I can raise a new issue for this other bug @goldmedal ?

**Comment by goldmedal** [comment]
> 
> This gets rewritten in the `GenericDialect` to:
> 
> ```
> SELECT ta.j1_id FROM (SELECT ta.j1_id FROM j1 AS ta)
> ```
> 
It's valid for DataFusion
```
> create table j1(j1_id int);
0 row(s) fetched. 
Elapsed 0.085 seconds.

> SELECT ta.j1_id FROM (SELECT ta.j1_id FROM j1 AS ta);
+-------+
| j1_id |
+-------+
+-------+
0 row(s) fetched. 
Elapsed 0.042 seconds.
```
Generally, `GenericDialect` fits the DataFusion syntax. This behavior makes sense to me.

> Which is invalid, because the subquery is un-aliased. I tested it in DuckDB to confirm:
> 
> ```
> D SELECT ta.j1_id FROM (SELECT ta.j1_id FROM j1 ta);
> Binder Error: Referenced table "ta" not found!
> Candidate tables: "unnamed_subquery"
> LINE 1: SELECT ta.j1_id FROM (SELECT ta.j1_id FROM j1 ...
> ```
> 
> Let's merge my PR, and I can raise a new issue for this other bug @goldmedal ?

How about `MySQLDialect`? It enables `requires_derived_table_alias`. Which would be the result?
```
SELECT ta.j1_id FROM (SELECT ta.j1_id FROM j1 ta) derived_projection
```
or
```
SELECT j1_id FROM (SELECT ta.j1_id FROM j1 ta) `derived_projection`
```

If it's the first one, I think it's an issue because it isn't valid. We can file an issue for it.
If it's the second one, I think the user can customize the `dialect::requires_derived_table_alias` for the DuckDB syntax purpose.

**Comment by peasee** [comment]
> If it's the first one, I think it's an issue because it isn't valid. We can file an issue for it.

My bad, I should've clarified better. It is the first one with `MySqlDialect`. I had just used `GenericDialect` as the example, but it affects both:

```
SELECT `ta`.`j1_id` FROM (SELECT `ta`.`j1_id` FROM `j1` AS `ta`) AS `derived_projection`
```

**Comment by goldmedal** [comment]
> > If it's the first one, I think it's an issue because it isn't valid. We can file an issue for it.
> 
> My bad, I should've clarified better. It is the first one with `MySqlDialect`. I had just used `GenericDialect` as the example, but it affects both:
> 
> ```
> SELECT `ta`.`j1_id` FROM (SELECT `ta`.`j1_id` FROM `j1` AS `ta`) AS `derived_projection`
> ```

I see. I think it's an issue but we can fix it in the follow-up PR. Let's merge this PR first.
Could you help to file an issue to trace it?

**Comment by peasee** [comment]
> I see. I think it's an issue but we can fix it in the follow-up PR. Let's merge this PR first. Could you help to file an issue to trace it?

https://github.com/apache/datafusion/issues/13027

**Comment by goldmedal** [comment]
Thanks @peasee and @phillipleblanc @alamb for reviewing. 👍

**Pull request merged**
